File: Q9.py
# 로프
# 탐욕법
# 백준 2217번

N = int(input())
limit = [int(input()) for _ in range(N)]

index = 0
startIdx = 0
weights = [] # 병렬로 연결된 로프들이 견딜 수 있는 무게가 저장
while index != N:
    # print(f'index = {index}')
    for i in range(startIdx, len(limit)):
        rope = [] # 선택한 로프를 담을 배열, index가 증가하고 다시 들어올 때 초기화
        for j in range(startIdx, i + 1):
            # print(f'limit[j] = {limit[j]}')
            rope.append(limit[j])
        
        rope.sort()
        result = rope[0] * len(rope) # 병렬로 견딜 수 있는 무게
        weights.append(result) # 최대 무게를 구하기 위해 배열에 담아둔다.

    startIdx += 1
    index += 1
weights.sort()
# print(weights)
print(weights[-1])

File: Q9.js
// Clear

let index = 0;
const limit = [10, 15, 150, 30];
let weights = [];

while(index != limit.length) {
    let cnt = 0;
    for (let i = index; i < limit.length; i++) {
      let weight = [];
      console.log(`줄의 갯수 = ${++cnt}`);
      for (let j = index; j <= i; j++) {
        weight.push(limit[j]);
      }
      // 들 수 있는 최대 무게 = 무게 안에서 가장 작은 값 * 무게 배열 갯수
      console.log(`로프들이 각각 견딜 수 있는 무게 : [${weight}]`);
      const answer = Math.min.apply(null, weight) * weight.length;
      weights.push(answer);
      console.log(`병렬로 견딜 수 있는 무게 = ${answer}`);
      console.log("============");
    }
    index++;
}

console.log(`정답 : ${Math.max.apply(null, weights)}`);